metricbeat/module/linux/perf/perf.go

```diff
@@ -0,0 +1,277 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// +build linux
+
+package perf
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/linux"
+
+	"github.com/hodgesds/perf-utils"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("linux", "perf", New)
+}
+
+type eventsConfig struct {
+	HardwareEvents bool `config:"hardware"`
+	SoftwareEvents bool `config:"software"`
+}
+
+type sampleConfig struct {
+	ProcessGlob string       `config:"process"`
+	Events      eventsConfig `config:"events"`
+}
+
+// Config holds the metricset config info for perf
+type Config struct {
+	SamplePeriod time.Duration  `config:"perf.sample_period" validate:"required"`
+	Processes    []sampleConfig `config:"perf.processes"`
+}
+
+// perfInfo is the "final" sampling data returned from a sample run
+// The struct contains metadata from the process along with sampling counters
+type perfInfo struct {
+	Metadata  common.MapStr
+	HwMetrics perf.HardwareProfile
+	SwMetrics perf.SoftwareProfile
+}
+
+// procInfo contains all the controlling information on a given PID.
+type procInfo struct {
+	PID          int
+	Metadata     common.MapStr
+	SoftwareProc perf.SoftwareProfiler
+	HardwareProc perf.HardwareProfiler
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	processes  []procInfo
+	period     time.Duration
+	continuous bool
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The linux perf metricset is beta.")
+	linuxModule, ok := base.Module().(*linux.Module)
+	if !ok {
+		return nil, errors.New("unexpected module type")
+	}
+	config := Config{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	if config.SamplePeriod > linuxModule.Period {
+		return nil, fmt.Errorf("Sample period of %s exceeds metricset period of %s", config.SamplePeriod, linuxModule.Period)
+	}
+
+	if config.SamplePeriod == 0 {
+		return nil, fmt.Errorf("Sample Period is zero")
+	}
+
+	var isContinuoius = false
+	if config.SamplePeriod == linuxModule.Period {
+		isContinuoius = true
+	}
```

Comment:
Nit: this could be shortened to `isContinuous := config.SamplePeriod == linuxModule.Period`